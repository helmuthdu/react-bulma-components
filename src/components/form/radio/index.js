import './radio.sass';

export * from './radio';
