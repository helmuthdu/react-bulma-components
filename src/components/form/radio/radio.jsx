import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';

const colors = Object.values(CONSTANTS.COLORS);

export const Radio = ({ className, style, disabled, value, children, checked, color, size, name, ...allProps }) => {
  const props = modifiers.clean(allProps);
  return (
    <label
      data-testid="radio-label"
      disabled={disabled}
      className={cn('b-radio radio', modifiers.getClassName(allProps), className)}
      style={style}
    >
      <input
        {...props}
        data-testid="radio-input"
        name={name}
        type="radio"
        value={value}
        disabled={disabled}
        checked={checked}
      />
      <span
        data-testid="radio-check"
        className={cn('check', {
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Radio.propTypes = {
  ...modifiers.propTypes,
  checked: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Radio.defaultProps = {
  ...modifiers.defaultProps,
  checked: false
};
