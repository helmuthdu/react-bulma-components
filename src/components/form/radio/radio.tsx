import cn from 'classnames';
import React from 'react';
import { Colors, Sizes } from '../../../constants';
import modifiers, { ElementModifier } from '../../../modifiers';

type RadioProps = ElementModifier & {
  color?: Colors;
  size?: Sizes;
  value?: string | number;
} & Omit<React.ComponentProps<'input'>, 'color' | 'size' | 'unselectable' | 'value'>;

export const Radio: React.FunctionComponent<RadioProps> = ({
  className,
  style,
  disabled,
  value,
  children,
  checked,
  color,
  size,
  name,
  testId,
  ...rest
}: RadioProps) => {
  const props = modifiers.clean(rest);
  return (
    <label
      data-testid={(Array.isArray(testId) ? testId[0] : testId) || undefined}
      className={cn('b-radio radio', modifiers.getClassName(rest), className)}
      style={style}
    >
      <input
        data-testid={Array.isArray(testId) ? testId[2] : undefined}
        name={name}
        type="radio"
        value={value}
        disabled={disabled}
        checked={checked}
        {...props}
      />
      <span
        data-testid={Array.isArray(testId) ? testId[1] : undefined}
        className={cn('check', {
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Radio.defaultProps = {
  ...modifiers.defaultProps,
  checked: false
};
