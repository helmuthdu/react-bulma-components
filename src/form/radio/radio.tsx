import clsx from 'clsx';
import * as React from 'react';
import { Colors, Sizes } from '../../constants';
import modifiers, { ElementModifier } from '../../modifiers';

type RadioProps = ElementModifier & {
  color?: Colors;
  size?: Sizes;
  value?: string | number;
} & Omit<React.ComponentProps<'input'>, 'ref' | 'color' | 'size' | 'unselectable' | 'value'>;

export const Radio = React.forwardRef<HTMLSelectElement, RadioProps>(
  ({ checked, children, className, color, disabled, name, size, style, value, ...props }, ref) => (
    <label className={clsx('radio-container radio', modifiers.getClassName(props), className)} style={style}>
      <input
        ref={ref}
        checked={checked}
        disabled={disabled}
        name={name}
        type="radio"
        value={value}
        {...modifiers.clean(props)}
      />
      <span
        className={clsx('check', {
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  )
);

Radio.defaultProps = {
  ...modifiers.defaultProps,
  checked: false
};
