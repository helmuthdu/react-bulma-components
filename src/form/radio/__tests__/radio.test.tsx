import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Radio } from '../radio';

describe('Radio component', () => {
  const testId = ['radio-label', 'radio-check', 'radio-input'];
  let onChange: jest.Mock;

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render', () => {
    const { asFragment } = render(
      <Radio onChange={onChange} name="test">
        Test <a>Give me</a>
      </Radio>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should concat classname in props with classname', () => {
    const { asFragment } = render(
      <Radio onChange={onChange} name="test" className="other-class this-is-a-test">
        Default
      </Radio>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should use inline styles', () => {
    const { asFragment } = render(
      <Radio onChange={onChange} name="test" style={{ height: 250 }}>
        Default
      </Radio>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be disabled, checked and with value', () => {
    const { asFragment } = render(
      <Radio name="test" value="TEST" onChange={onChange} checked disabled>
        Default
      </Radio>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should set input checked if checked', () => {
    const { getByTestId } = render(<Radio testId={testId} onChange={onChange} checked />);
    expect(getByTestId('radio-input')).toHaveAttribute('checked', '');
  });

  it('should change value on change event', () => {
    const { getByTestId } = render(
      <Radio testId={testId} onChange={onChange}>
        Text
      </Radio>
    );
    fireEvent.click(getByTestId('radio-label'));
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
