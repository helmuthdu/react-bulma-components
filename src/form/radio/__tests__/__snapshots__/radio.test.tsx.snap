// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component should be disabled, checked and with value 1`] = `
<DocumentFragment>
  <label
    class="radio-container radio"
  >
    <input
      checked=""
      disabled=""
      name="test"
      type="radio"
      value="TEST"
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Default
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio component should concat classname in props with classname 1`] = `
<DocumentFragment>
  <label
    class="radio-container radio other-class this-is-a-test"
  >
    <input
      name="test"
      type="radio"
      value=""
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Default
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio component should render 1`] = `
<DocumentFragment>
  <label
    class="radio-container radio"
  >
    <input
      name="test"
      type="radio"
      value=""
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Test 
      <a>
        Give me
      </a>
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio component should use inline styles 1`] = `
<DocumentFragment>
  <label
    class="radio-container radio"
    style="height: 250px;"
  >
    <input
      name="test"
      type="radio"
      value=""
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Default
    </span>
  </label>
</DocumentFragment>
`;
